### PR TITLE
Timing implementation

### PR DESCRIPTION
**Pull request by jrmadsen**
- Created C++ "auto_timer" and Python "auto_timer"
  - Start timer upon construction
  - Stop timer upon deconstructor
  - Python layer uses various utility functions to get unique tags
- All timers started in Python have prefix "[pyc]"
- All timers started in C++ have prefix "[cxx]"
- C++ timers are placed in the ctoast.cpp file
- All pipelines have been updated to use auto_timers

### CODE REVIEW

File: src/libtoast/util/toast_timing_manager.cpp [hunk 1]
```diff
@@ -62,14 +63,41 @@ toast::util::timer& toast::util::timing_manager::timer(const string_t& key)
...
     timer_pair_t _pair(key, m_timer_map[key]);
     m_timer_list.push_back(_pair);
 
     return m_timer_map[key];
 }
+
 //============================================================================//
 
 
+
```
Comment (keskitalo) [inline src/libtoast/util/toast_timing_manager.cpp]: There is a ton of whitespace at the end of this file.

File: src/python/timing.py [hunk 2]
```diff
@@ -3,27 +3,74 @@
...
+
+from os.path import dirname
+from os.path import basename
+from os.path import join
+
+#------------------------------------------------------------------------------#
+
+def FUNC(back = 1):
+    """Function that emulates __FUNCTION__ macro"""
+    ret = ("%s" % (sys._getframe(back).f_code.co_name))
```
Comment (keskitalo) [inline src/python/timing.py]: Old school string formatting?

**Comment by jrmadsen** [comment]
Ah yes. We can delete that whitespace at the end of the file. 

I already have a threshold set (see sex/libtoast/until/toast_base_timer.cpp. I think the function is called "output"). Basically the default precision of the timer is 3 decimal places (which in its current units, this is milliseconds) so anything less than 1/pow(10, precision) is omitted from the result.

The main timer will always appear first and it will set the width of the "time" string so even in larger runs it will be easy to gauge the magnitude of the time values (and thereby dismiss the irrelevant times) because all of the decimal places will always be exactly lined up vertically. Doing anything beyond that might be difficult given various run times and might introduce a potential for lost data. 

Feel free to remove any of the old timing. After basically spending an hour+ semi-mindlessly copy-and-pasting the "auto_timer" everywhere in Python, I couldn't mentally handle that task until later. 

Sent from my iPhone

--
Jonathan R. Madsen

**Comment by jrmadsen** [comment]
This is used by the python auto_timer to get the file that the timer was created in. It's one of many I implemented just so I didn't have to customize each insertion of an auto-timer. 

Sent from my iPhone

--
Jonathan R. Madsen

**Comment by tskisner** [comment]
The curly brace relocating in toast.cpp is just noise, but will be reverted anyway during the big format cleanup.  Please change the python string formatting to use format() instead of the percent sign, and then feel free to merge.

**Pull request merged**